README.md
Passerelle connector to communicate with ATAL
================================================

Installation
------------

 - add to Passerelle installed apps settings:
   INSTALLED_APPS += ('passerelle_imio_ia_tech',)

 - enable module:
   PASSERELLE_APP_PASSERELLE_IMIO_IA_TECH_ENABLED = True


Usage
-----

 - create and configure new connector
   - Title/description: whatever you want
   - Certificate check: uncheck if the service has no valid certificate

 - test service by clicking on the available links
   - the /testConnection/ endpoint try to establish a connection with ATAL
   - the /test_createItem/ endpoint try to create a new point in ATAL



Usage in w.c.s.
---------------



CHANGES.rst
Changelog
=========

0.4.4
------------------

- [TELE-913] pass requester contact info [dmu]


0.4.3
------------------

- split fpm command on multiple line
- build package with python3
- do not set auto dependencies
[nhi]

0.4.2
------------------

- [TELE-933] add dont_write_bytecode (fpm include pyc files in production fix) [dmu]


0.4.1
------------------

- [TELE-933] only clean on success [dmu] (2nd)
  fix merge code commited by mistake
  https://support.imio.be/browse/TELE-933

0.4.0
------------------

- [TELE-933] only clean on success [dmu] 
  https://support.imio.be/browse/TELE-933


0.3.9
------------------

- [TELE-933] add cleanWs() to fix workspace not cleaned [dmu] 
  https://support.imio.be/browse/TELE-933


0.3.8
------------------

- [TELE-930] add endpoint to retrieve details of a work request [dmu]
  https://support.imio.be/browse/TELE-930

0.3.7
------------------

- [TELE-927] add comments to models enpoints [dmu]
  https://support.imio.be/browse/TELE-927

0.3.6
------------------

- [TELE-927] add attachment post to ATAL endpoint [dmu]
  https://support.imio.be/browse/TELE-927


0.3.5
------------------

- [TELE-913] add work request post [dmu]



0.3.4
------------------

- [TELE-913] add 0001_initial migration file [dmu]


0.3.3
------------------

- [TELE-913] fix test endpoint url [dmu]


0.3.2
------------------

- [TELE-913] remove unused import and field + pre-commit and pyproject optim [dmu]


0.3.1
------------------

- [TELE-913] return test endpoint response and remode duplicate changes.rst

0.3.0
------------------

- [TELE-913] start over for atal 6 [dmu]

0.2.1f
------------------

- [TELE-695] use passerelle json_load to avoid encoding issues [dmu]

0.2.1e
------------------

- [TELE-642] fix encoding again in soap.py

0.2.1d
------------------

- [TELE-638] In models, changes implementation of StringIO to be python3 complient
- [TELE-640] Fix Test method by adding request as parameter
- [TELE-641] Use BytesIO instead of StringIO

0.2.1c
------------------

- [TELE-636] fix parenthesis position

0.2.1b
------------------

- [TELE-636] Fix json bug as it loads a bytes instead of a str since python3

0.2.1a
------------------

- Adapt Jenkinsfile to install package python3/dist-package instead of python2
[nhislaire] [dmuyshond]


passerelle_imio_ia_tech/soap.py
# Passerelle - uniform access to data and services
# Copyright (C) 2015  Entr'ouvert
#
# This program is free software: you can redistribute it and/or modify it
# under the terms of the GNU Affero General Public License as published
# by the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Affero General Public License for more details.
#
# You should have received a copy of the GNU Affero General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

# borrowed from https://pypi.python.org/pypi/suds_requests
# and https://docs.oracle.com/cd/E50245_01/E50253/html/vmprg-soap-example-authentication-python.html

from io import BytesIO


from suds.client import Client
from suds.plugin import DocumentPlugin, MessagePlugin
from suds.transport import Reply
from suds.transport.http import HttpAuthenticated


class Filter(MessagePlugin):
    def marshalled(self, context):
        context.envelope.set("xmlns:xm", "http://www.w3.org/2005/05/xmlmime")

    def received(self, context):
        reply = context.reply
        context.reply = reply[reply.find("<?xml version"): reply.rfind(">") + 1]


class Handlewsdl(DocumentPlugin):
    def loaded(self, context):
        # unknown types, so present them as strings
        context.document = context.document.replace(
            'type="iph:DossierID"', 'type="xsd:string"'
        )
        context.document = context.document.replace(
            'type="iph:TypeTechnique"', 'type="xsd:string"'
        )


class Transport(HttpAuthenticated):
    def __init__(self, model, **kwargs):
        self.model = model
        HttpAuthenticated.__init__(self, **kwargs)

    def get_requests_kwargs(self):
        kwargs = {}
        if self.model.username:
            kwargs["auth"] = (self.model.username, self.model.password)
        if self.model.keystore:
            kwargs["cert"] = (self.model.keystore.path, self.model.keystore.path)
        if not self.model.verify_cert:
            kwargs["verify"] = False
        return kwargs

    def open(self, request):
        # only use our custom handler to fetch service resources, not schemas
        # from other namespaces
        if "www.w3.org" in request.url:
            return HttpAuthenticated.open(self, request)
        resp = self.model.requests.get(
            request.url, headers=request.headers, **self.get_requests_kwargs()
        )
        return BytesIO(resp.content)

    def send(self, request):
        request.message = request.message.replace(b'contentType', b'xm:contentType')
        self.addcredentials(request)
        resp = self.model.requests.post(
            request.url,
            data=request.message,
            headers=request.headers,
            **self.get_requests_kwargs()
        )
        return Reply(resp.status_code, resp.headers, resp.content)


# plugins=[Handlewsdl(), Filter()], plugins=[ImportDoctor(imp)]
def get_client(instance):
    transport = Transport(instance)
    return Client(
        instance.wsdl_url, plugins=[], transport=transport, cache=None, autoblend=True
    )


passerelle_imio_ia_tech/migrations/0001_initial.py
# -*- coding: utf-8 -*-
# Generated by Django 1.11.29 on 2021-06-14 14:29
from __future__ import unicode_literals

from django.db import migrations, models


class Migration(migrations.Migration):

    initial = True

    dependencies = [
        ('base', '0029_auto_20210202_1627'),
    ]

    operations = [
        migrations.CreateModel(
            name='imio_atal',
            fields=[
                (
                    'id',
                    models.AutoField(
                        auto_created=True,
                        primary_key=True,
                        serialize=False,
                        verbose_name='ID',
                    ),
                ),
                ('title', models.CharField(max_length=50, verbose_name='Title')),
                ('slug', models.SlugField(unique=True, verbose_name='Identifier')),
                ('description', models.TextField(verbose_name='Description')),
                (
                    'base_url',
                    models.URLField(
                        help_text='API base URL',
                        max_length=256,
                        verbose_name='Base URL',
                    ),
                ),
                ('api_key', models.CharField(max_length=128, verbose_name='API Key')),
                (
                    'users',
                    models.ManyToManyField(
                        blank=True,
                        related_name='_imio_atal_users_+',
                        related_query_name='+',
                        to='base.ApiUser',
                    ),
                ),
            ],
            options={
                'verbose_name': 'Connecteur Atal (iMio)',
            },
        ),
    ]


passerelle_imio_ia_tech/models.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

import base64
from io import BytesIO
import requests

from django.db import models
from django.utils.translation import ugettext_lazy as _
from django.utils.six.moves.urllib_parse import urljoin
from passerelle.compat import json_loads
from passerelle.base.models import BaseResource
from passerelle.utils.api import endpoint
from passerelle.utils.jsonresponse import APIError


class imio_atal(BaseResource):
    base_url = models.URLField(
        max_length=256,
        blank=False,
        verbose_name=_("Base URL"),
        help_text=_("API base URL"),
    )
    api_key = models.CharField(max_length=128, verbose_name=_("API Key"))

    category = _("Business Process Connectors")

    class Meta:
        verbose_name = "Connecteur Atal (iMio)"

    @classmethod
    def get_verbose_name(cls):
        return cls._meta.verbose_name

    @endpoint(perm='can_access', description=_('Test methods'))
    def Test(self, request):
        atal_response = requests.get(
            url=self.base_url + "api/Test",
            headers={"Accept": "text/plain", "X-API-Key": self.api_key},
        )
        atal_response_format = "{} - {}".format(
            atal_response.status_code, atal_response.text
        )
        return atal_response_format

    @endpoint(
        perm='can_access',
        description=_('Post '),
        methods=['post'],
        parameters={
            'atal_requester_id': {
                'description': "Numéro de l'utilisateur demandeur dans ATAL",
                'type': 'integer',
            },
            'atal_contact_firstname': {
                'description': "Prénom du contact transmis vers ATAL",
                'type': 'string',
            },
            'atal_contact_lastname': {
                'description': "Nom du contact transmis vers ATAL",
                'type': 'string',
            },
            'atal_contact_email': {
                'description': "Adresse mail du contact transmis vers ATAL",
                'type': 'string',
            },
            'atal_contact_phone': {
                'description': "Numéro de téléphone du contact transmis vers ATAL",
                'type': 'string',
            },
            'atal_contact_mobile': {
                'description': "Numéro de mobile du contact transmis vers ATAL",
                'type': 'string',
            },
            'atal_contact_address': {
                'description': "Adresse du contact transmis vers ATAL",
                'type': 'string',
            },
            'atal_contact_zipcode': {
                'description': "Code postal du contact transmis vers ATAL",
                'type': 'string',
            },
            'atal_contact_city': {
                'description': "Localité du contact transmis vers ATAL",
                'type': 'string',
            },
            "atal_operator": {
                'description': "Nom de l'opérateur",
                'type': "string",
            },
            "atal_object": {
                'description': "Objet de la demande",
                'type': "string",
            },
            "atal_description": {
                'description': "Description de la demande",
                'type': "string",
            },
            "atal_desired_date": {
                'description': "Date désirée pour la demande (ex: 2021-06-16T13:30:25.365Z)",
                'type': "string",
            },
            "atal_recipient_id": {
                'description': "Numéro identifiant du destinataire",
                'type': "integer",
            },
            "atal_requesting_department_id": {
                'description': "Numéro identifiant du département demandeur",
                'type': "integer",
            },
            "atal_request_type": {
                'description': "Numéro identifiant du type de requête",
                'type': "integer",
            },
            "atal_request_date": {
                'description': "Date de la requête",
                'type': "string",
            },
            "atal_localization": {
                'description': "Lieu de l'intervention",
                'type': "string",
            },
            "atal_patrimony_id": {
                'description': "Numéro identifiant du patrimoine",
                'type': "integer",
            },
            "atal_longitude": {
                'description': "Coordonée longitude du lieu de l'intervention",
                'type': "integer",
            },
            "atal_latitude": {
                'description': "Coordonée latitude du lieu de l'intervention",
                'type': "integer",
            },
        },
    )
    def create_work_request(self, request, *args, **kwargs):
        data_from_publik = json_loads(request.body)  # http data from wcs webservice
        # commented parameters below are not required
        data_to_atal = {
            "Contact": {
                "FirstName": data_from_publik["atal_contact_firstname"],
                "LastName": data_from_publik["atal_contact_lastname"],
                "Email": data_from_publik["atal_contact_email"],
                "Phone": data_from_publik["atal_contact_phone"],
                "Mobile": data_from_publik["atal_contact_mobile"],
                "Address1": data_from_publik["atal_contact_address"],
                "ZipCode": data_from_publik["atal_contact_zipcode"],
                "City": data_from_publik["atal_contact_city"],
            },
            "RequesterId": data_from_publik["atal_requester_id"],
            # "Operator": data_from_publik["atal_operator"],
            "Object": data_from_publik["atal_object"],
            "Description": data_from_publik["atal_description"],
            # "DesiredDate": data_from_publik["atal_desired_date"],
            "RecipientId": data_from_publik["atal_recipient_id"],
            "RequestingDepartmentId": data_from_publik["atal_requesting_department_id"],
            "RequestType": data_from_publik["atal_request_type"],
            # "RequestDate": data_from_publik["atal_request_date"],
            "Localization": data_from_publik["atal_localization"],
            # "PatrimonyId": data_from_publik["atal_patrimony_id"],
            "Longitude": data_from_publik["atal_longitude"],
            "Latitude": data_from_publik["atal_latitude"],
        }
        # Raw data that works locally with Postman or Python request
        """
        data_to_atal = {
            "RequesterId": 13,
            "RecipientId": 1004,
            "RequestingDepartmentId": 690,
            "Localization": "Bruxelles",
            "latitude": 50.844998072856825,
            "longitude": 4.349986346839887,
            "Object": "Test via Python request (objet)",
            "Description": "Test via Python request (description)",
            "RequestType": 1001,
        }
        """
        response = self.requests.post(
            urljoin(
                self.base_url,
                '/api/WorksRequests',  # Endpoints are described in ATAl Swagger
            ),
            json=data_to_atal,
            headers={
                "accept": "application/json",
                # X-API-KEY is visible in ATAL admin panel
                # and set in the passerelle connector settings.
                "X-API-Key": self.api_key,
                "Content-Type": "application/json",
            },
        )
        # Error handling (WIP)
        """
        if response.raise_for_status():
            error_desc = {'message': response.text}
            raise APIError(
                '{} (HTTP error {})'.format(response.text, response.status_code),
                data=error_desc,
            )
        else:
        """
        response_json = response.json()
        return {'data': response_json}

    @endpoint(
        perm='can_access',
        description=_('Post files and join them to and ATAL work request.'),
        methods=['post'],
        parameters={
            "atal_work_request_uuid": {
                "description": "Numéro d'identification uuid de la demande dans ATAL.",
                "type": "string",
                # "example_value": "11865d87-336b-49b3-e181-08d76f93d6b4",
            },
            "atal_attachment1": {
                "description": "Fichier 1",
            },
        },
    )
    def post_attachment(self, request, *args, **kwargs):
        data_from_publik = json_loads(request.body)  # http data from wcs webservice

        # ATAL 6 require a multipart/formdata request
        # with a bytes encoded file. That's why we use
        # BytesIO here to convert Base64 image from wcs to bytes
        decoded_image = base64.b64decode(
            data_from_publik["atal_attachment1"]["content"]
        )
        image_for_atal = BytesIO(decoded_image)

        response = self.requests.post(
            urljoin(
                self.base_url,
                '/api/WorksRequests/'
                # uuid is fetched from ATAL work request creation response in wcs vars
                + data_from_publik["atal_work_request_uuid"] + '/Attachments',
            ),
            headers={
                # X-API-KEY is visible in ATAL admin panel
                # and set in the passerelle connector settings.
                "X-API-Key": self.api_key,
            },
            # This is multipart/formdata http request file syntax
            files={
                'file': (
                    data_from_publik["atal_attachment1"]["filename"],
                    image_for_atal,
                    data_from_publik["atal_attachment1"]["content_type"],
                )
            },
        )

        # Return http request response text if there is an error
        # or http response converted to json if there is not
        if response.raise_for_status():
            error_desc = {'message': response.text}
            raise APIError(
                '{} (HTTP error {})'.format(response.text, response.status_code),
                data=error_desc,
            )
        else:
            response = response.json()

        return {'data': response}  # must return dict

    @endpoint(
        perm='can_access',
        description=_('Return work request details from ATAL.'),
        methods=['post'],
        parameters={
            "atal_work_request_uuid": {
                "description": "Numéro d'identification uuid de la demande dans ATAL.",
                "type": "string",
                "example_value": "11865d87-336b-49b3-e181-08d76f93d6b4",
            },
        },
    )
    def get_work_request_details(self, request, *args, **kwargs):
        data_from_publik = json_loads(request.body)  # http data from wcs webservice

        # ATAL 6 require uuid in the url as an endpoint
        # for example :
        # https://demov6.imio-app.be/api/WorksRequests/11865d87-336b-49b3-e181-08d76f93d6b4

        work_request_uuid = data_from_publik["atal_work_request_uuid"]

        response = self.requests.get(
            urljoin(
                self.base_url,
                '/api/WorksRequests/'
                # uuid is fetched from ATAL work request creation response in wcs vars
                + data_from_publik["atal_work_request_uuid"],
            ),
            headers={
                "accept": "application/json",
                # X-API-KEY is visible in ATAL admin panel
                # and set in the passerelle connector settings.
                "X-API-Key": self.api_key,
            },
        )

        # Return http request response text if there is an error
        # or http response converted to json if there is not
        if response.raise_for_status():
            error_desc = {'message': response.text}
            raise APIError(
                '{} (HTTP error {})'.format(response.text, response.status_code),
                data=error_desc,
            )
        else:
            response = response.json()

        return {'data': response}  # must return dict
